### DISCUSSION TITLE
upgrade to php 8.1

### DISCUSSION THREAD

**Discussion by steamboatid** [MyIntervals/PHP-CSS-Parser#380] [Show and tell]
dear all,

i manage to upgrade [sabberworm/PHP-CSS-Parser](https://github.com/sabberworm/PHP-CSS-Parser) to php 8.1
you may find my repo at [steamboatid/PHP-CSS-Parser](https://github.com/steamboatid/PHP-CSS-Parser)

i did 4 phase modifications :

phase 1 :
- create dedicated lxc with php 7.4.30.
- fresh clone main repo, install all packages needed, do phpunit test.
- create phpstan baseline
- run phpunit test again, to make sure all test passed
- do simple refactor using rector, just to insert arguments type at functions.
- exec phpstan and fix all of phpstan reports
- exec composer fix:php 
- exec phpunit test until all test passed

phase 2 :
- upgrade phpunit to phpunit 7.1.5
- exec phpstan and fix all of phpstan reports
- exec composer fix:php 
- exec phpunit test until all test passed

phase 3 :
- create lxc with php 8.1.7 from sury.org
- exec phpstan and fix all of phpstan reports
- exec composer fix:php 
- exec phpunit test until all test passed

phase 4 :
- upgrade phpunit to 8.5.27
- exec phpunit test until all test passed

precautions :
- it may not compatible with php below 8.1
- i only test with with phpunit and not yet tested with real applications

**Comment by oliverklee** [comment]
Hi @steamboatid, thanks for digging into this!

Concerning the minimum supported PHP version this library needs to support, we had a discussion about this roughly one year ago: https://github.com/sabberworm/PHP-CSS-Parser/issues/251

Basically, it boils down to "we need to support the lowest PHP version that still has [at least 5% usage share with WordPress](https://wordpress.org/about/stats/)" (because this library is used by WordPress).

As of today, PHP 5.6 still has a share of 5.97%, and is (very) slowly going down. As PHP 7.0 and 7.1 both have a very low share, we can bump the minimum PHP version to 7.2 hopefully sometime soon.

**Comment by oliverklee** [comment]
We have now raised the PHP requirement to >= 7.2 and support PHP up to 8.3 (as covered by tests). Closing.
